src/backend/distributed/utils/shardinterval_utils.c

```diff
@@ -242,25 +242,63 @@ FindShardInterval(Datum partitionColumnValue, ShardInterval **shardIntervalCache
 				  int shardCount, char partitionMethod, FmgrInfo *compareFunction,
 				  FmgrInfo *hashFunction, bool useBinarySearch)
 {
-	ShardInterval *shardInterval = NULL;
+	Datum searchedValue = partitionColumnValue;
+	int shardIndex = INVALID_SHARD_INDEX;
+
+	if (partitionMethod == DISTRIBUTE_BY_HASH)
+	{
+		searchedValue = FunctionCall1(hashFunction, partitionColumnValue);
+	}
+
+	shardIndex = FindShardIntervalIndex(searchedValue, shardIntervalCache,
+										shardCount, partitionMethod, compareFunction,
+										hashFunction, useBinarySearch);
+
+	if (shardIndex == INVALID_SHARD_INDEX)
+	{
+		return NULL;
+	}
+
+	return shardIntervalCache[shardIndex];
+}
+
+
+/*
+ * FindShardIntervalIndex finds the index of the shard interval which covers
+ * the searched value. Note that the searched value must be the hashed value
+ * of the original value if the distribution method is hash.
+ */
+static int
+FindShardIntervalIndex(Datum searchedValue, ShardInterval **shardIntervalCache,
+					   int shardCount, char partitionMethod, FmgrInfo *compareFunction,
+					   FmgrInfo *hashFunction, bool useBinarySearch)
+{
+	int shardIndex = INVALID_SHARD_INDEX;
 
 	if (partitionMethod == DISTRIBUTE_BY_HASH)
 	{
-		int hashedValue = DatumGetInt32(FunctionCall1(hashFunction,
-													  partitionColumnValue));
 		if (useBinarySearch)
 		{
 			Assert(compareFunction != NULL);
 
-			shardInterval = SearchCachedShardInterval(Int32GetDatum(hashedValue),
-													  shardIntervalCache, shardCount,
-													  compareFunction);
+			shardIndex = SearchCachedShardInterval(searchedValue, shardIntervalCache,
+												   shardCount, compareFunction);
+
+			/* we should always return a valid shard index for hash partitioned tables */
+			if (shardIndex == INVALID_SHARD_INDEX)
+			{
+				ereport(ERROR, (errcode(ERRCODE_DATA_EXCEPTION),
+								errmsg("cannot find shard interval"),
+								errdetail("The searched value is not in the range "
```

Comment:
I'm not the best person to write error messages. But, one alternative:  `Hash of the partition column value does not fall into a single shard`